tests/tests_fabric/strategies/test_model_parallel_integration.py

```diff
@@ -0,0 +1,489 @@
+        assert all(not torch.equal(batches_dp[0], batches_dp[i]) for i in range(1, len(batches_dp)))
+
+        output = model(batch)
+        fabric.backward(output.sum())
+        optimizer.step()
+        optimizer.zero_grad()
+
+
+@RunIf(min_torch="2.3", min_cuda_gpus=4, standalone=True)
+@pytest.mark.parametrize(
+    "precision",
+    [
+        pytest.param(
+            "16-mixed", marks=pytest.mark.xfail(reason="Precision plugin does not implement ShardedGradScaler yet")
+        ),
+        pytest.param("bf16-mixed", marks=RunIf(bf16_cuda=True)),
+    ],
+)
+def test_train_save_load(precision, tmp_path):
+    """Test 2D-parallel training, saving and loading precision settings."""
+    strategy = ModelParallelStrategy(
+        _parallelize_feed_forward_fsdp2_tp,
+        data_parallel_size=2,
+        tensor_parallel_size=2,
+    )
+    fabric = Fabric(accelerator="cuda", devices=4, strategy=strategy, precision=precision)
+    fabric.launch()
+
+    fabric.seed_everything(0)
+    with fabric.init_module(empty_init=True):
+        model = FeedForward()
+    model = fabric.setup(model)
+    optimizer = torch.optim.AdamW(model.parameters())
+    optimizer = fabric.setup_optimizers(optimizer)
+    output = model(torch.rand(2, 32, device=fabric.device))
+    fabric.backward(output.sum())
+    optimizer.step()
+    optimizer.zero_grad()
+
+    checkpoint_path = fabric.broadcast(str(tmp_path / "dist-checkpoint"))
+
+    params_before = [p.full_tensor().clone() for p in model.parameters()]
+    state = {"model": model, "optimizer": optimizer, "steps": 1}
+    fabric.save(checkpoint_path, state)
+    assert set(os.listdir(checkpoint_path)) == {
+        ".metadata",
+        "__0_0.distcp",
+        "__1_0.distcp",
+        "__2_0.distcp",
+        "__3_0.distcp",
+    }
+
+    # re-init all objects and resume
+    strategy = ModelParallelStrategy(
+        _parallelize_feed_forward_fsdp2_tp,
+        data_parallel_size=2,
+        tensor_parallel_size=2,
+    )
+    fabric = Fabric(accelerator="cuda", devices=4, strategy=strategy, precision=precision)
+    fabric.launch()
+
+    fabric.seed_everything(0)
+    with fabric.init_module(empty_init=True):
+        model = FeedForward()
+    model = fabric.setup(model)
+    optimizer = torch.optim.AdamW(model.parameters())
+    optimizer = fabric.setup_optimizers(optimizer)
+    output = model(torch.rand(2, 32, device=fabric.device))
+    fabric.backward(output.sum())
+    optimizer.step()
+    optimizer.zero_grad()
+
+    # check correctness with loaded state
+    state = {"model": model, "optimizer": optimizer, "steps": 0}
+    metadata = fabric.load(checkpoint_path, state)
+    for p0, p1 in zip(params_before, (p.full_tensor() for p in model.parameters())):
+        torch.testing.assert_close(p0, p1, atol=0, rtol=0, equal_nan=True)
+
+    # check user data in state reloaded
+    # TODO: This should be 1, torch.distributed.checkpoint only supports tensor data
```

Comment:
Non ShardedTensor or DTensor objects are assumed to be replicated (same value in all ranks).

Reply:
This will come when I finish implementing the checkpointing. For now, it will only work for tensors.